client/Windows/wf_defaults.c

```diff
@@ -0,0 +1,100 @@
+#include <Windows.h>
+#include <wincred.h>
+#include <stdio.h>
+#include <malloc.h>
+#include <freerdp/settings.h>
+
+#pragma warning(disable : 4706 4094) // assignment within conditional expression
+
+#define FreeRDP_ServerHostname (20)
+#define FreeRDP_Username (21)
+#define FreeRDP_Password (22)
+#define FreeRDP_GatewayHostname (1986)
+#define FreeRDP_GatewayUsername (1987)
+#define FreeRDP_GatewayPassword (1988)
+
+typedef struct rdp_settings rdpSettings;
+
+PCWSTR ValidateString(const BYTE* pb, ULONG cb)
+{
+	if (!pb || !cb)
+		return 0;
+
+	if (cb & (sizeof(WCHAR) - 1))
+		return 0;
+
+	if (*(WCHAR*)(pb + cb - sizeof(WCHAR)))
+		return 0;
+
+	return (PCWSTR)pb;
+}
+
+void AddDefaultSettings_I(rdpSettings* settings, size_t idHostname, size_t idUsername,
+                          size_t idPassword)
+{
+	PCSTR ServerHostname = freerdp_settings_get_string(settings, idHostname);
+
+	if (!ServerHostname)
+		return;
+
+	BOOL bExistUserName = freerdp_settings_get_string(settings, idUsername) != 0;
+	BOOL bExistPassword = freerdp_settings_get_string(settings, idPassword) != 0;
+
+	if (bExistUserName && bExistPassword)
+		return;
+
+	PCREDENTIALA Credential;
+	PCWSTR lpWideCharStr;
+
+	PSTR Password;
+	PSTR UserName;
+
+	PSTR TargetName = 0;
+	int len = _vsnprintf(TargetName, 0, "TERMSRV/%s", (va_list)&ServerHostname);
+	len++;
+	TargetName = (PSTR)malloc(len);
+	_vsnprintf(TargetName, len, "TERMSRV/%s", (va_list)&ServerHostname);
+
+	if (!TargetName)
+		return;
+
+	TargetName[len - 1] = 0;
+
+	if (!CredReadA(TargetName, CRED_TYPE_GENERIC, 0, &Credential))
+		return;
+
+	if (!bExistPassword)
+	{
+		ULONG cch = Credential->CredentialBlobSize;
+
+		if (lpWideCharStr = ValidateString(Credential->CredentialBlob, cch))
+		{
+			Password = 0, len = 0, cch /= sizeof(WCHAR);
+
+			len = WideCharToMultiByte(CP_UTF8, 0, lpWideCharStr, cch, Password, len, 0, 0);
+			Password = (PSTR)malloc(len);
+			WideCharToMultiByte(CP_UTF8, 0, lpWideCharStr, cch, Password, len, 0, 0);
+			
+			if (Password)
+			{
+				freerdp_settings_set_string(settings, idPassword, Password);
+			}
+		}
+	}
+
+	if (!bExistUserName)
+	{
+		if (UserName = Credential->UserName)
+		{
+			freerdp_settings_set_string(settings, idUsername, UserName);
+		}
+	}
+
+	CredFree(Credential);
+}
+
+EXTERN_C void WINAPI AddDefaultSettings(_Inout_ rdpSettings* settings)
```

Comment:
No need to export the function, only used in the binary itself (not in an external API)
but creating a header file for that (which is included where it is used) is highly recommended. (that way the compiler knows properly of which type arguments and return are)